File: LeetCode/combination_sum.cpp
class Solution {
public:
    vector< vector<int> > res;
    
    void solve(vector<int>& candidates, int start, vector<int> current, int s, int target) {
        if(s > target)
            return;
        
        if(s == target) {
            res.push_back(current);
            return;
        }
        
        for(int i = start; i < candidates.size(); i++) { // Define the start point, excluding the ones already picked
            current.push_back(candidates[i]);
            solve(candidates, i, current, s + candidates[i], target);
            current.pop_back();
        }
    }
    
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        vector<int> current;
        
        solve(candidates, 0, current, 0, target);
        
        return res;
    }
};

File: LeetCode/combination_sum.go
func solve(candidates, curr []int, target, start int, ans *[][]int) {
    if target < 0 {
        return
    }

    if target == 0 {
        tmp := make([]int, len(curr))
        copy(tmp, curr)
        *ans = append(*ans, tmp)
        return
    }

    for i := start; i < len(candidates); i++ {
        curr = append(curr, candidates[i])
        solve(candidates, curr, target - candidates[i], i, ans)
        curr = curr[:len(curr)-1]
    }
}

func combinationSum(candidates []int, target int) [][]int {
    ans := [][]int{} 

    solve(candidates, []int{}, target, 0, &ans)

    return ans
}
